examples/fft_batch_speeds.py

```diff
@@ -1,91 +1,149 @@
-  but will be limited by your computer's RAM.
-- Here, batch size is the size of the 3D stack in z.
-- Note that each pipeline runs 4 FFTs. For example, batch 8 runs 8*4=32 FFTs.
+.. admonition:: Notes on warmup runs and PyFFTW
+
+    Each FFT Filter has a warmup run to reduce outliers.
+    This is especially important for PyFFTW. Always keep a single batch size
+    when using PyFFTW, as during the first run at each batch size,
+    PyFFTW will find the fastest way to perform the FFT.
 
 """
 import time
 import matplotlib.pylab as plt
 import numpy as np
 import qpretrieve
 from qpretrieve.data_array_layout import convert_data_to_3d_array_layout
-from qpretrieve.fourier import FFTFilterNumpy, FFTFilterPyFFTW
+from qpretrieve.fourier import FFTFilterNumpy, FFTFilterPyFFTW, FFTFilterCupy
 
 # load the experimental data
 edata = np.load("./data/hologram_cell.npz")
 
 n_transforms_list = [8, 16, 32, 64, 128, 256]
 subtract_mean = True
 padding = True
-# we take the PyFFTW speeds from the second run
-fft_interfaces = [FFTFilterNumpy, FFTFilterPyFFTW, FFTFilterPyFFTW]
+
+# the first run is always used as a warmup
+fft_interfaces = [FFTFilterNumpy, FFTFilterNumpy]
+if FFTFilterCupy is not None:
+    fft_interfaces.extend([FFTFilterPyFFTW, FFTFilterPyFFTW])
+if FFTFilterCupy is not None:
+    fft_interfaces.extend([FFTFilterCupy, FFTFilterCupy])
+
 filter_name = "disk"
 filter_size = 1 / 2
-speed_norms = {}
 
 # load and prep the data
-data_2d = edata["data"].copy()
-data_2d_bg = edata["bg_data"].copy()
-data_3d_prep, _ = convert_data_to_3d_array_layout(data_2d)
-data_3d_bg_prep, _ = convert_data_to_3d_array_layout(data_2d_bg)
+data_2d, data_2d_bg = edata["data"].copy(), edata["bg_data"].copy()
+input_data_3d, _ = convert_data_to_3d_array_layout(data_2d)
+input_data_bg_3d, _ = convert_data_to_3d_array_layout(data_2d_bg)
 
+speed_batch_norms, speed_ffts = {}, {}
+print("Each FFTFilter will run twice, as a warmup run is "
+      "required for comparison.")
 for fft_interface in fft_interfaces:
-    results = {}
+    results_batch, results_fft = {}, {}
     for n_transforms in n_transforms_list:
         print(f"Running {n_transforms} transforms with "
               f"{fft_interface.__name__}")
 
         # create batches
-        data_3d = np.repeat(data_3d_prep, repeats=n_transforms, axis=0)
-        data_3d_bg = np.repeat(data_3d_bg_prep, repeats=n_transforms, axis=0)
+        data_3d = np.repeat(input_data_3d, repeats=n_transforms, axis=0)
+        data_3d_bg = np.repeat(input_data_bg_3d, repeats=n_transforms, axis=0)
 
-        assert data_3d.shape == data_3d_bg.shape == (n_transforms,
-                                                     edata["data"].shape[0],
-                                                     edata["data"].shape[1])
+        assert data_3d.shape == data_3d_bg.shape == (
+            n_transforms, edata["data"].shape[0], edata["data"].shape[1])
 
         t0 = time.time()
```

Comment:
Use `time.perf_counter()`. `time.time()` is not accurate (especially on Windows).